### PR TITLE
Add primitive Value Type reference projection support (part 1)

### PR DESCRIPTION
**Pull request by hangshao0**
1. There are 2 ram classes generated from a primitive value type. One is the L type ram class and the other is the Q type ram class. Add fields Ltype and Qtype into J9Class. No non-primitive value classes, these 2 fields points to the J9Class itself.
2. For primitive VT, only the Q type can be instanciated. Embed the L type J9Class header in the Q type. The different behaviours of L type and Q type can be controlled by J9Class->classFlags.
3. Add new flag J9_FINDCLASS_FLAG_QTYPE to indicate if it is a Q type that is to be found or generated.

issue #14858

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

File: runtime/oti/j9.h [hunk 1]
```diff
@@ -323,11 +323,14 @@ static const struct { \
...
 #define J9_CLASS_ALLOWS_LOCKING(clazz) J9_ARE_NO_BITS_SET((clazz)->classFlags, J9_CLASS_DISALLOWS_LOCKING_FLAGS)
 #define J9_IS_J9CLASS_VALUEBASED(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsValueBased)
+#define J9_IS_J9CLASS_GENERATED(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassGenerated)
 #ifdef J9VM_OPT_VALHALLA_VALUE_TYPES
 #define J9CLASS_UNPADDED_INSTANCE_SIZE(clazz) J9_VALUETYPE_FLATTENED_SIZE(clazz)
 #define J9_IS_J9CLASS_VALUETYPE(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsValueType)
 #define J9_IS_J9CLASS_PRIMITIVE_VALUETYPE(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsPrimitiveValueType)
+#define J9_IS_J9CLASS_GENERATED_VT_LTYPE(clazz) (J9_IS_J9CLASS_VALUETYPE(clazz) && J9_IS_J9CLASS_GENERATED(clazz))
 #define J9_IS_J9CLASS_FLATTENED(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassIsFlattened)
+#define J9_ARE_J9CLASSES_SAME(c1, c2) (((c1) == (c2)) || ((c2) == (c1)->Ltype) || ((c2) == (c1)->Qtype))
```
Comment (gacholio) [inline runtime/oti/j9.h:333]: This name isn't great - can you find a more descriptive term instead of "same"?

**Comment by gacholio** [comment]
Are Q and L types equivalent for purposes of checkcast/instanceof and the associated java methods?

File: runtime/oti/j9.h (hunk 1 above)
Comment (gacholio) [reply runtime/oti/j9.h:333]: That seems reasonable. @tajila what do you think?

**Comment by gacholio** [comment]
Is "equivalent" a better term than "companion" then?

**Comment by hangshao0** [comment]
> Is "equivalent" a better term than "companion" then?

Yes, I agree. "equivalent" is more obvious than "companion" on what the macro means.

File: runtime/oti/j9.h (hunk 1 above)
Comment (gacholio) [reply runtime/oti/j9.h:333]: See discussion below.

Comment (hangshao0) [reply runtime/oti/j9.h]: Updated.

Comment (tajila) [reply runtime/oti/j9.h]: Equivalent makes sense

**Comment by hangshao0** [comment]
This is no longer needed.

**Pull request closed without merging**